### PR TITLE
Made home work add Category entity and modified Book entity.

### PR DESCRIPTION
**Pull request by SergeyGordeychuk**

### CODE REVIEW

File: src/main/java/onlinebookstore/mapper/BookMapper.java [hunk 1]
```diff
@@ -1,17 +1,35 @@
...
+    @AfterMapping
+    default void setCategories(@MappingTarget BookDto bookDto, Book book) {
+        bookDto.setCategories(mapToIds(book.getCategories()));
+    }
+
+    default Set<Long> mapToIds(Set<Category> categories) {
+        Set<Long> categoryIds = new HashSet<>();
+        for (Category category : categories) {
+            categoryIds.add(category.getId());
+        }
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/mapper/BookMapper.java]: use stream api

File: src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public CategoryDto update(Long id, CategoryRequestDto requestDto) {
+        Category category = categoryMapper.toEntity(requestDto);
+        category.setId(id);
+        if (categoryRepository.existsById(id)) {
+            categoryRepository.save(category);
+        } else {
+            throw new NoSuchElementException("Can't update category: " + category);
+        }
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java]: Invert IF condition and check it at the very beginning of the method

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/onlinebookstore/service/category/CategoryService.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import onlinebookstore.dto.category.CategoryRequestDto;
+import onlinebookstore.dto.category.CategoryDto;
+import org.springframework.data.domain.Pageable;
+
+public interface CategoryService {
+    List<CategoryDto> findAll(Pageable pageable);
+    CategoryDto getById(Long id);
+    CategoryDto save(CategoryRequestDto requestDto);
+    CategoryDto update(Long id, CategoryRequestDto requestDto);
+    void delete(Long id);
```
Comment (MAMentorFX) [inline src/main/java/onlinebookstore/service/category/CategoryService.java]: empty lines between methods

File: src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+@RequiredArgsConstructor
+@Service
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable)
+                .stream()
```
Comment (MAMentorFX) [inline src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java]:
```suggestion
        return categoryRepository.findAll(pageable).stream()

```

File: src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public CategoryDto update(Long id, CategoryRequestDto requestDto) {
+        if (!categoryRepository.existsById(id)) {
+            throw new NoSuchElementException("Can't find category by id: " + id );
+        }
+        Category category = categoryMapper.toEntity(requestDto);
+        category.setId(id);
+        categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
```
Comment (MAMentorFX) [inline src/main/java/onlinebookstore/service/category/CategoryServiceImpl.java]: isolate it with transaction

File: src/main/resources/db/changelog/changes/11-create-books-categories-table.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            referencedTableName: books
+            referencedColumnNames: id
+            onDelete: CASCADE
+        - addForeignKeyConstraint:
+            baseTableName: books_categories
+            baseColumnNames: category_id
+            constraintName: fk_books_categories_category_id
+            referencedTableName: categories
+            referencedColumnNames: id
+            onDelete: CASCADE
```
Comment (MAMentorFX) [inline src/main/resources/db/changelog/changes/11-create-books-categories-table.yaml]:
```suggestion
            onDelete: CASCADE
           
```

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/onlinebookstore/dto/category/CategoryRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package onlinebookstore.dto.category;
+
+import jakarta.validation.constraints.NotNull;
+
+public record CategoryRequestDto(
+        @NotNull
```
Comment (koval-kn) [inline src/main/java/onlinebookstore/dto/category/CategoryRequestDto.java:6]: its better
```suggestion
        @NotBlank
```

**Review by koval-kn — APPROVED** [review]

**Pull request merged**